awsshell/selectmenu.py

```diff
@@ -0,0 +1,264 @@
+        padding = ' ' * (self.width - len(option))
+        return [(token, ' %s%s ' % (option, padding))]
+
+
+def create_select_menu_layout(msg, menu_control,
+                              show_meta=False,
+                              reserve_space_for_menu=True):
+    """Construct a layout for the given message and menu control."""
+    def get_prompt_tokens(_):
+        return [(Token.Prompt, msg)]
+
+    # Ensures that the menu has enough room to display it's options
+    def get_prompt_height(cli):
+        if reserve_space_for_menu and not cli.is_done:
+            return LayoutDimension(min=8)
+        else:
+            return LayoutDimension()
+
+    input_processors = [DefaultPrompt(get_prompt_tokens)]
+    prompt_layout = FloatContainer(
+        HSplit([
+            Window(
+                BufferControl(input_processors=input_processors),
+                get_height=get_prompt_height
+            )
+        ]),
+        [
+            Float(
+                # Display the prompt starting below the cursor
+                left=len(msg),
+                ycursor=True,
+                content=ConditionalContainer(
+                    content=Window(
+                        content=menu_control,
+                        # only display 1 - 7 lines of completions
+                        height=LayoutDimension(min=1, max=7),
+                        # display a scroll bar
+                        scroll_offsets=ScrollOffsets(top=0, bottom=0),
+                        right_margins=[ScrollbarMargin()],
+                        # don't make the menu wider than the options
+                        dont_extend_width=True
+                    ),
+                    # Only display the prompt while the buffer is relevant
+                    filter=~IsDone()
+                )
+            )
+        ]
+    )
+
+    # Show meta information with the buffer isn't done and show_meta is True
+    meta_filter = ~IsDone() & to_simple_filter(show_meta)
+    return HSplit([
+        prompt_layout,
+        ConditionalContainer(
+            filter=meta_filter,
+            content=Window(
+                height=LayoutDimension.exact(1),
+                content=FillControl(u'\u2500', token=Token.Line)
+            )
+        ),
+        ConditionalContainer(
+            filter=meta_filter,
+            content=Window(
+                # Meta information takes up at most 15 lines
+                height=LayoutDimension(max=15),
+                content=BufferControl(
+                    buffer_name='INFO',
+                    # TODO discuss pygments import voodoo
+                    lexer=PygmentsLexer(find_lexer_class('JSON'))
+                )
+            ),
+        )
+    ])
+
+
+class SelectMenuApplication(Application):
+    """Wrap Application, providing the correct layout, keybindings, etc."""
+
+    def __init__(self, message, options, *args, **kwargs):
+        self.menu_control = SelectMenuControl(options)
```

Comment:
This is a really long init. You should break out the logic into a helper function.